projects/core/directives/dropdown/dropdown-options.directive.ts

```diff
@@ -0,0 +1,102 @@
+import {
+    Directive,
+    FactoryProvider,
+    forwardRef,
+    Inject,
+    InjectionToken,
+    Input,
+    Optional,
+    SkipSelf,
+} from '@angular/core';
+import {tuiDefaultProp} from '@taiga-ui/cdk';
+import {
+    TuiDropdownWidth,
+    TuiHorizontalDirection,
+    TuiVerticalDirection,
+} from '@taiga-ui/core/types';
+
+export interface TuiDropdownOptions {
+    readonly align: TuiHorizontalDirection;
+    readonly direction: TuiVerticalDirection | null;
+    readonly limitWidth: TuiDropdownWidth;
+    readonly minHeight: number;
+    readonly maxHeight: number;
+}
+
+/** Default values for dropdown options */
+export const TUI_DROPDOWN_DEFAULT_OPTIONS: TuiDropdownOptions = {
+    align: `left`,
+    direction: null,
+    limitWidth: `auto`,
+    maxHeight: 400,
+    minHeight: 80,
+};
+
+export const TUI_DROPDOWN_OPTIONS = new InjectionToken<TuiDropdownOptions>(
+    `[TUI_DROPDOWN_OPTIONS] Default parameters for dropdown directive`,
+    {
+        factory: () => TUI_DROPDOWN_DEFAULT_OPTIONS,
+    },
+);
+
+export const tuiDropdownOptionsProvider: (
+    options: Partial<TuiDropdownOptions>,
+) => FactoryProvider = (override: Partial<TuiDropdownOptions>) => ({
+    provide: TUI_DROPDOWN_OPTIONS,
+    deps: [
+        [new Optional(), TuiDropdownOptionsDirective],
+        [new SkipSelf(), TUI_DROPDOWN_OPTIONS],
+    ],
+    useFactory: (
+        directive: TuiDropdownOptions | null,
+        options: TuiDropdownOptions,
+    ): TuiDropdownOptions => {
+        const defaults = options === TUI_DROPDOWN_DEFAULT_OPTIONS;
+        const fallback = defaults ? {...options} : options;
+        const result = directive || fallback;
+
+        Object.keys(override).forEach(key => {
+            // Update props with new defaults before inputs are processed
+            (result as any)[key] = override[key as keyof TuiDropdownOptions];
```

Comment:
why not

```
(result as unknown as TuiDropdownOptions)
```

Reply:
Because the interface is all `readonly` and will not allow us to mutate it here. Besides I believe TypeScript does not understand that we match keys and will not be able to determine that type of `override[key]` is the same as `result[key]`.